### PR TITLE
🌱 test: add PreWaitForControlplaneToBeUpgraded to ClusterUpgradeConformanceSpec

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Implements additional checks to ensure the cluster is operational during an update.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

/area e2e-testing

### CODE REVIEW

File: test/e2e/cluster_upgrade_test.go [hunk 1]
```diff
@@ -57,7 +70,8 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass [ClusterC
 	})
 })
 
-var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA control plane [ClusterClass]", func() {
+var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA control plane [ClusterClass]", Label("Foo"), func() {
```
Comment (fabriziopandini) [inline test/e2e/cluster_upgrade_test.go:73]:
```suggestion
var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA control plane [ClusterClass]", func() {
```

File: test/e2e/cluster_upgrade_test.go [hunk 2]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+				cluster := &clusterv1.Cluster{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: workloadClusterNamespace,
+						Name:      workloadClusterName,
+					},
+				}
+				Expect(managementClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(cluster), cluster)).To(Succeed())
+
+				// This replaces the WaitForControlPlaneMachinesToBeUpgraded function and additionally:
+				// * checks that kube-proxy is healthy
```
Comment (fabriziopandini) [inline test/e2e/cluster_upgrade_test.go:103]: What about adding a not explaining that we are doing this test in order to ensure that non static pods remain healthy on CP machines during upgrade

File: test/e2e/cluster_upgrade_test.go [hunk 3]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+
+					// Collect information about:
+					// * how many control-plane machines already got upgraded
+					// * control-plane machines which are in deletion but waiting for the pre-drain hook
+					// * workload cluster nodes
+					// * kube-proxy pods
+					var upgraded int64
+					deletingMachinesWithPreDrainHook := []clusterv1.Machine{}
+					for _, m := range machines {
+						if *m.Spec.Version == cluster.Spec.Topology.Version && conditions.IsTrue(&m, clusterv1.MachineNodeHealthyCondition) {
```
Comment (fabriziopandini) [inline test/e2e/cluster_upgrade_test.go:120]: q: why are we checking clusterv1.MachineNodeHealthyCondition? as far as I remember it only checks for a dummy condition to not exists, so if I'm not wrong it doesn't really give added value 🤔

File: test/e2e/cluster_upgrade_test.go [hunk 4]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+					for _, node := range nodes.Items {
+						for _, condition := range node.Status.Conditions {
+							if condition.Type != corev1.NodeReady || condition.Status == corev1.ConditionTrue {
+								continue
+							}
+							errList = append(errList, errors.Errorf("Node's %s Ready condition is false", node.GetName()))
+						}
+					}
+
+					// Check if the expected number of kube-proxy pods exist and all of them are healthy.
```
Comment (fabriziopandini) [inline test/e2e/cluster_upgrade_test.go:151]: might be let's add a not specifying that we are checking kube proxy both on old and new CP nodes, as well as on workers (across the entire cluster)

File: test/e2e/cluster_upgrade_test.go [hunk 5]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+							}
+							errList = append(errList, errors.Errorf("Pod's %s Ready condition is false", pod.GetName()))
+						}
+					}
+
+					if err := kerrors.NewAggregate(errList); err != nil {
+						return 0, errors.Wrap(err, "blocking upgrade because cluster is not stable")
+					}
+
+					// Remove pre-drain webhook from machines because all current machines are considered ok.
```
Comment (fabriziopandini) [inline test/e2e/cluster_upgrade_test.go:168]:
```suggestion
					// At this stage all current machines are considered ok, so remove the pre-drain webhook from a CP to unblock the next step of the upgrade
```

**Review by fabriziopandini — COMMENTED** [review]
Thanks for implementing this test! just a few nits from my side

File: test/e2e/cluster_upgrade_test.go (hunk 3 above)
Comment (sbueringer) [reply test/e2e/cluster_upgrade_test.go:120]: Wait what. For a dummy condition to not exist?

Comment (chrischdi) [reply test/e2e/cluster_upgrade_test.go:120]: jep, I was expecting this to check this as mentioned in the comment

File: test/e2e/cluster_upgrade_test.go [hunk 6]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+						}
+						if !m.DeletionTimestamp.IsZero() && m.Annotations[preDrainHook] == "true" {
+							deletingMachinesWithPreDrainHook = append(deletingMachinesWithPreDrainHook, m)
+						}
+					}
+
+					wlClient := managementClusterProxy.GetWorkloadCluster(ctx, workloadClusterNamespace, workloadClusterName).GetClient()
+					nodes := corev1.NodeList{}
+					if err := wlClient.List(ctx, &nodes); err != nil {
+						return 0, errors.New("failed to list nodes in workload cluster")
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:131]: Why are we hiding the error here and below? (instead of wrapping it)

File: test/e2e/cluster_upgrade_test.go [hunk 7]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+					if err := wlClient.List(ctx, &kubeProxyPods, client.InNamespace(metav1.NamespaceSystem), client.MatchingLabels{"k8s-app": "kube-proxy"}); err != nil {
+						return 0, errors.New("failed to list kube-proxy pods in workload cluster")
+					}
+
+					errList := []error{}
+
+					// Check all nodes to be Ready.
+					for _, node := range nodes.Items {
+						for _, condition := range node.Status.Conditions {
+							if condition.Type != corev1.NodeReady || condition.Status == corev1.ConditionTrue {
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:144]: Would be easier to read if we check for the case we're trying to catch

\+  let's add the condition message to the error

File: test/e2e/cluster_upgrade_test.go [hunk 8]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+								continue
+							}
+							errList = append(errList, errors.Errorf("Node's %s Ready condition is false", node.GetName()))
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:147]: Based on the current check we only know it's not true, not if it's false (but I think we should also catch the unknown case, the message just doesn't reflect that)

Comment (sbueringer) [reply test/e2e/cluster_upgrade_test.go:147]: Same for kube-proxy

File: test/e2e/cluster_upgrade_test.go [hunk 9]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+					if len(deletingMachinesWithPreDrainHook) > 0 {
+						if len(deletingMachinesWithPreDrainHook) > 1 {
+							return 0, errors.Errorf("expected a maximum of 1 control-plane machines to be in deleting and having the pre-drain hook but got %d", len(deletingMachinesWithPreDrainHook))
+						}
+
+						// Removing pre-drain hook from machine.
+						m := &deletingMachinesWithPreDrainHook[0]
+						patchHelper, err := patch.NewHelper(m, managementClusterProxy.GetClient())
+						if err != nil {
+							return 0, errors.Wrapf(err, "creating patchHelper for Machine %s", klog.KObj(m))
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:178]:
```suggestion
							return 0, err
```
No need, patch helper already builds that message

File: test/e2e/cluster_upgrade_test.go [hunk 10]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+						}
+						delete(m.Annotations, preDrainHook)
+
+						if err := patchHelper.Patch(ctx, m); err != nil {
+							return 0, errors.Wrapf(err, "failed to remove pre-drain annotation from Machine %s", klog.KObj(m))
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:183]: Same here

File: test/e2e/cluster_upgrade_test.go [hunk 11]
```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+						// Return to enter the function again.
+						return 0, errors.Errorf("deletion of Machine %s was blocked by pre-drain hook", klog.KObj(m))
+					}
+
+					if int64(len(machines)) > upgraded {
+						return 0, errors.New("old Machines remain")
+					}
+
+					return upgraded, nil
+				}, e2eConfig.GetIntervals("k8s-upgrade-and-conformance", "wait-machine-upgrade")...).Should(Equal(controlPlaneMachineCount), "Timed out waiting for all control-plane machines in Cluster %s to be upgraded to kubernetes version %s", klog.KObj(cluster), cluster.Spec.Topology.Version)
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:195]: Does it work correctly to check for only a number while returning a number and an error?

Comment (sbueringer) [reply test/e2e/cluster_upgrade_test.go:195]: I think I would just pass in g Gomega into this eventually and then check everything (and not return anything)

Comment (sbueringer) [reply test/e2e/cluster_upgrade_test.go:195]: Hm but not sure if that works with the retry logic you need

File: test/e2e/cluster_upgrade_test.go (hunk 3 above)
Comment (fabriziopandini) [reply test/e2e/cluster_upgrade_test.go:120]: This is how we are configuring MHC in E2E tests:

```
    machineHealthCheck:
      maxUnhealthy: 100%
      unhealthyConditions:
        - type: e2e.remediation.condition
          status: "False"
          timeout: 20s
```

So (in E2E tests only) MHC is testing for a dummy `e2e.remediation.condition`, not for NodeReady, NodeMemoryPressure etc

Comment (sbueringer) [reply test/e2e/cluster_upgrade_test.go:120]: But this is the MachineNodeHealthyCondition not the MachineHealthCheckSucceeded condition

Comment (sbueringer) [reply test/e2e/cluster_upgrade_test.go:120]: It is set to true here: https://github.com/kubernetes-sigs/cluster-api/blob/d7db259b7de2e113c1658f5f547ff70bef9c9120/internal/controllers/machine/machine_controller_noderef.go#L160

This should have nothing to do with MHC's

File: test/e2e/cluster_upgrade_test.go (hunk 11 above)
Comment (chrischdi) [reply test/e2e/cluster_upgrade_test.go:195]: its the same pattern we already use in WaitForControlPlaneMachinesToBeUpgraded

Comment (chrischdi) [reply test/e2e/cluster_upgrade_test.go:195]:
```
If multiple values are returned by the function, Eventually will pass the first value to the matcher and require that all others are zero-valued.
```
From the godoc

File: test/e2e/cluster_upgrade_test.go (hunk 6 above)
Comment (chrischdi) [reply test/e2e/cluster_upgrade_test.go:131]: dunno :D forgot to wrap. Changing to error wrapping.

**Comment by chrischdi** [comment]
/test pull-cluster-api-e2e-main

rebase

File: test/e2e/cluster_upgrade_test.go (hunk 11 above)
Comment (sbueringer) [reply test/e2e/cluster_upgrade_test.go:193]: Alright, sounds good!

Comment (sbueringer) [reply test/e2e/cluster_upgrade_test.go:193]: Didn't remember

File: test/e2e/cluster_upgrade_test.go [hunk 12]
```diff
@@ -69,9 +83,115 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: workloadClusterNamespace,
+						Name:      workloadClusterName,
+					},
+				}
+				Expect(managementClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(cluster), cluster)).To(Succeed())
+
+				// This replaces the WaitForControlPlaneMachinesToBeUpgraded function to ensure that
+				// all static Pods and kube-proxy as well as all Nodes are healthy during the whole upgrade
+				// across all Nodes of the Cluster. For doing this it uses a pre-drain hook.
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:104]:
```suggestion

				// This replaces the WaitForControlPlaneMachinesToBeUpgraded function to verify via a pre-drain hook 
				// that all static Pods, kube-proxy and all Nodes are becoming healthy before we let the upgrade 
				// process precede by removing the pre-drain hook.
				// This captures cases where static Pods, kube-proxy and all Nodes would only become healthy after 
				// all control plane Machines have been upgraded.
```
I know it's just a nuance but I think it's an important one

File: test/e2e/cluster_upgrade_test.go [hunk 13]
```diff
@@ -69,9 +83,115 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+						ClusterName: cluster.Name,
+						Namespace:   cluster.Namespace,
+					})
+
+					// Collect information about:
+					// * how many control-plane machines already got upgraded
+					// * control-plane machines which are in deletion but waiting for the pre-drain hook
+					// * workload cluster nodes
+					// * kube-proxy pods
+					var upgraded int64
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:117]:
```suggestion
					var upgradedAndHealthy int64
```
nit Maybe, up to you

File: test/e2e/cluster_upgrade_test.go [hunk 14]
```diff
@@ -69,9 +83,115 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+						return 0, errors.Wrapf(err, "failed to list kube-proxy pods in workload cluster")
+					}
+
+					errList := []error{}
+
+					// Check all nodes to be Ready.
+					for _, node := range nodes.Items {
+						for _, condition := range node.Status.Conditions {
+							if condition.Type == corev1.NodeReady && condition.Status != corev1.ConditionTrue {
+								errList = append(errList, errors.Errorf("Expected the Ready condition for Node %s to be true but got %s instead: %s", node.GetName(), condition.Status, condition.Message))
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:145]: nit: error strings should be lower case, same in l.159

File: test/e2e/cluster_upgrade_test.go [hunk 15]
```diff
@@ -69,9 +83,115 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+							}
+						}
+					}
+
+					// Check if the expected number of kube-proxy pods exist and all of them are healthy for all existing Nodes of the Cluster.
+					if len(nodes.Items) != len(kubeProxyPods.Items) {
+						errList = append(errList, errors.Errorf("exected %d kube-proxy pods to exist, got %d", len(nodes.Items), len(kubeProxyPods.Items)))
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:152]:
```suggestion
						errList = append(errList, errors.Errorf("expected %d kube-proxy pods to exist, got %d", len(nodes.Items), len(kubeProxyPods.Items)))
```
nit

File: test/e2e/cluster_upgrade_test.go [hunk 16]
```diff
@@ -69,9 +83,115 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+						return 0, errors.Wrap(err, "blocking upgrade because cluster is not stable")
+					}
+
+					// At this stage all current machines are considered ok, so remove the pre-drain webhook from a CP to unblock the next step of the upgrade.
+					if len(deletingMachinesWithPreDrainHook) > 0 {
+						if len(deletingMachinesWithPreDrainHook) > 1 {
+							return 0, errors.Errorf("expected a maximum of 1 control-plane machines to be in deleting and having the pre-drain hook but got %d", len(deletingMachinesWithPreDrainHook))
+						}
+
+						// Removing pre-drain hook from machine.
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:172]: Should we only do this if there are no Machines with deletionTimestamp and without pre-drain hook?

(just to ensure we're not hitting strange edge cases where we are removing pre-drain hooks from multiple machines)

File: test/e2e/cluster_upgrade_test.go [hunk 17]
```diff
@@ -69,9 +83,122 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+				Eventually(func() (int64, error) {
+					machines := framework.GetControlPlaneMachinesByCluster(ctx, framework.GetControlPlaneMachinesByClusterInput{
+						Lister:      managementClusterProxy.GetClient(),
+						ClusterName: cluster.Name,
+						Namespace:   cluster.Namespace,
+					})
+
+					// Collect information about:
+					// * how many control-plane machines already got upgradedAndHealthy
+					// * control-plane machines which are in deletion but waiting for the pre-drain hook
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:116]:
```suggestion
					// * control-plane machines which are in deletion
```

File: test/e2e/cluster_upgrade_test.go [hunk 18]
```diff
@@ -69,9 +83,122 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
...
+					// * workload cluster nodes
+					// * kube-proxy pods
+					var upgradedAndHealthy int64
+					deletingMachines := []clusterv1.Machine{}
+					for _, m := range machines {
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:121]: @chrischdi Can you please open a separate PR to fix the description of the MachineNodeHealthyCondition?

```go
	// MachineNodeHealthyCondition provides info about the operational state of the Kubernetes node hosted on the machine by summarizing  node conditions.
	// If the conditions defined in a Kubernetes node (i.e., NodeReady, NodeMemoryPressure, NodeDiskPressure, NodePIDPressure, and NodeNetworkUnavailable) are in a healthy state, it will be set to True.
	MachineNodeHealthyCondition ConditionType = "NodeHealthy"
```
It's at least wrong because we never look at NodeNetworkUnavailable (didn't check if NodeNetworkUnavailable even exists)

**Review by sbueringer — COMMENTED** [review]
Last nits from my side

**Comment by sbueringer** [comment]
/assign @fabriziopandini 
(for a final review)

File: test/e2e/cluster_upgrade_test.go (hunk 18 above)
Comment (chrischdi) [reply test/e2e/cluster_upgrade_test.go:121]: https://github.com/kubernetes-sigs/cluster-api/pull/11178

**Comment by sbueringer** [comment]
Thx!!

Really nice improvement

/lgtm

**Review by fabriziopandini — COMMENTED** [review]
Great work!
/lgtm
/approve

**Comment by sbueringer** [comment]
This additional validation found this issue: https://github.com/kubernetes-sigs/cluster-api/issues/11296

Let's also add it to release-1.8

**Comment by sbueringer** [comment]
I'll do a manual cherry-pick
